File: python/rafter_cli/scanners/git_diff_scan.py
import os

from ..scanners.regex_scanner import RegexScanner, ScanResult
from ..utils.git_diff import AddedDiffLine


def scan_added_diff_lines(
    added_lines: list[AddedDiffLine],
    repo_root: str,
    custom_patterns=None,
) -> list[ScanResult]:
    """Scan parsed git diff added lines with the patterns engine."""
    if not added_lines:
        return []

    scanner = RegexScanner(custom_patterns)
    by_file: dict[str, list] = {}

    for entry in added_lines:
        abs_path = os.path.join(repo_root, entry.file)
        matches = scanner.scan_line(entry.text, entry.line)
        if matches:
            by_file.setdefault(abs_path, []).extend(matches)

    return [ScanResult(file=f, matches=m) for f, m in by_file.items() if m]

File: node/src/scanners/git-diff-scan.ts
import path from "path";
import { AddedDiffLine } from "../utils/git-diff.js";
import { RegexScanner, ScanResult } from "./regex-scanner.js";

/**
 * Scan only added/modified lines from a parsed git diff (+ side).
 * Uses the patterns engine — git line scope is incompatible with betterleaks'
 * whole-file `dir` scan (same as the PreToolUse staged hook).
 */
export function scanAddedDiffLines(
  addedLines: AddedDiffLine[],
  repoRoot: string,
  customPatterns?: Array<{ name: string; regex: string; severity: string }>,
): ScanResult[] {
  if (addedLines.length === 0) return [];

  const scanner = new RegexScanner(customPatterns);
  const byFile = new Map<string, ScanResult["matches"]>();

  for (const { file, line, text } of addedLines) {
    const absPath = path.resolve(repoRoot, file);
    const matches = scanner.scanLine(text, line);
    if (matches.length === 0) continue;
    const existing = byFile.get(absPath) ?? [];
    existing.push(...matches);
    byFile.set(absPath, existing);
  }

  return [...byFile.entries()].map(([file, matches]) => ({ file, matches }));
}
